### ISSUE TITLE
Work around translocation by updating into /Applications

### ISSUE THREAD

**Issue by pornel** [sparkle-project/Sparkle#1012]
It might be a silly idea, but how about writing updates to /Applications instead of current app's path, if the app is run from a translocated folder (or a DMG)?

Some apps suggest copying themselves to /Applications. We could kill two birds with one stone by doing copy+update.

**Comment by zorgiepoo** [comment]
> if the app is run from a translocated folder (or a DMG)?

The appropriate detection is if the app is running inside a read-only volume.

Honestly I think doing nothing is better than writing somewhere else potentially confusing the user, and dealing with edge cases involved in that. Also developers should be abiding to follow Apple's guidelines, or Apple should be improving the opting-out of translocation experience, not us trying to work around it IMO.

**Comment by pornel** [comment]
> Apple should be improving the opting-out of translocation experience

I wouldn't bet on that. I worry that Apple's sees non-App-Store software only as a liability (potentially malicious, legacy software), so they're not interested in making that experience any better.

**Comment by pornel** [comment]
I would like to improve translocation experience, because I've ran into #923, but I absolutely don't want to touch the DMG format ever.

**Comment by zorgiepoo** [comment]
> I wouldn't bet on that. I worry that Apple sees non-App-Store software only as a liability (potentially malicious, legacy software), so they're not interested in making that experience any better.

I didn't say they *would*. I said if a change would be to made, they *should* be the ones doing it. They still care about downloading applications outside of the store, else they wouldn't have introduced translocation.

> I would like to improve translocation experience, because I've ran into #923, but I absolutely don't want to touch the DMG format ever.

I think fighting against the grain of Apple is not a wise decision, especially not from a popular software update framework, which will directly impact everyone else involved. I can see a potential future where gatekeeper does not bode well for apps that aren't distributed in signed archives. If that future becomes a reality, you'll have no choice but to use dmg. I'm not keen on short-term bandage workarounds.

For what it's worth, I also discussed [this issue](https://github.com/Squirrel/Squirrel.Mac/issues/182) in another known macOS updater library.

**Comment by pornel** [comment]
> I'm not keen on short-term bandage workarounds.

What would be the long term solution? 

BTW, DMGs, signed or not, are also affected (app run off the image can't be updated).

Something like that exists: https://github.com/potionfactory/LetsMove/

but in Sparkle we already have all the components for detecting read-only volumes, reliable copying, authorization, etc. 

We could improve the "read-only volume, error!" alert by making it "read-only volume, fix it" alert.

**Comment by zorgiepoo** [comment]
One note: we don't have reliable secure authorization yet, and neither does LetsMove. I saw @jakepetroules merged the recent [docs PR](https://github.com/sparkle-project/sparkle-project.github.io/pull/40); I'm not comfortable with recommending LetsMove officially. (As an aside, in general, please give a chance for a pull request to be reviewed by all before merging changes; same could be said with closing newly reported issues)

Thinking about it a bit, I think the challenges involved in moving an app into /Applications/ [securely](https://github.com/sparkle-project/Sparkle/issues/363#issue-36945601) could be harder and different than replacing an existing bundle securely. I wouldn't be tricked into thinking the problem is that relatable.

(By the way, I personally like running applications inside from dmg's at least temporarily for trial-ing purposes, without being notified of a prompt)

If Apple somehow makes the installation experience better for non-MAS apps, that would be long term, I guess. Short term hacks would be us trying to clear a quarantine bit, or installing the app elsewhere on disk, or moving the app like what LetsMove does. The issues I have with those kind of have to do with future compatibility, user confusion / disruption, or increasing complexity of the framework and maintaining code that I think is outside of Sparkle's scope. (I think Sparkle's scope is to securely update an existing bundle in-place).

**Comment by zorgiepoo** [comment]
To expand on authorization and security (in general, not specific to Sparkle) I think it helps to imagine a scenario where you have a continuously running tool running at the highest privileges, and a client running at the lowest privileges. A client can ask it to perform a task that it was built for, but the tool can't trust the client. This isn't what my fork mirrors exactly, but this is the model that I don't want to stray away from.

In our case, a client can ask the tool to update/replace a bundle that was signed by the same developer. It will only update bundles if the update that is found has been signed in a matching way. The worst case is that the tool is allowed to update any Sparkle supported bundle. Additional code signing requirements can be added to ensure the tool and the bundle it's updating have been signed by the same entity, but this involves complications from a framework point of view, and at other times we may even like the flexibility of updating any other bundle.

Now, you can try to introduce a change that says updates for bundles that are on read-only mounts should be installed in /Applications. But remember, the client can't be trusted. This allows an untrusted client to be able to install a Sparkle supported app they like into /Applications/ when it was not there initially (Btw, an "edge case": what if an application with the same name already exists in /Applications - it could be a copy of the same app, or a different app!)

In any case, I think a workaround should *at least* be deferred until Sparkle has a more proper authorization / security model. And that's only if we think a workaround should be done.

**Comment by pornel** [comment]
Totally agreed on the sandboxed client/daemon model.

Copying, like all Sparkle activity, should be deferred to at least second run. It could be deferred until there's an update available, so it wouldn't interrupt trying out apps any more than regular updates.

If we only write new (verified + upgraded) app to /Applications, then it avoids risk of propagating any potentially-tampered apps. Copying of course should never overwrite existing app in /Applications.

**Comment by zorgiepoo** [comment]
With the theoretical model above my point is that it allows a client to install any Sparkle app (even if verified and updated securely) in /Applications once a copying policy is introduced.

**Comment by zorgiepoo** [comment]
>  Copying of course should never overwrite existing app in /Applications

Translocation makes this less obvious. An interesting case is what if the app being translocated came from /Applications (and translocation itself is not reliably detectable)

An obvious case of user confusion is the old app in a users dock no longer working, and a duplicated set of icons.

**Comment by pornel** [comment]
> With the theoretical model above my point is that it allows a client to install any Sparkle app (even if "verified" and updated securely) in /Applications

Is it under assumption that there would be a privileged daemon already running? Otherwise the malicious app would have to ask for authorization, and then from attacker's perspective getting sparkle to ask for auth isn't any better than making the malicious app to ask.

But if we assume there is a privileged daemon, then perhaps we could make the daemon open alert window asking for permission to copy (or even auth again), so that malicious apps can't abuse it.

**Comment by zorgiepoo** [comment]
> Is it under assumption that there would be a privileged daemon already running?

Yeah, which isn't how my fork currently works btw, but is something I kept thinking about not breaking too far away from. But you can think about it this way instead: it's under the assumption that the daemon can't trust the input from the other side. Currently this means the daemon can be tricked into updating some other Sparkle based app in-place, which considering as the worst possibility, I think that's alright.. With a copying policy, that means the daemon can be tricked into installing any Sparkle based app into /Applications. Unless you have the daemon ensure it is code signed the same way as the bundle being updated, but now that's an extra requirement to add, and sometimes we don't want that restriction.

Daemons don't show user interface by the way, and communication should be possible through IPC. Also the identity of the other side of the connection can't be ensured.

Food for thought anyway. I was just trying to expand on how the copying problem may not be *that* similar to Sparkle currently. Ignoring that there are other UX issues trying to work around translocation.

**Comment by pornel** [comment]
So how about helping "DIY" copying instead?

1. Show /Applications in Finder
2. Show the app in Finder
3. Quit the app

**Comment by zorgiepoo** [comment]
Not certain that will work if the app is translocated (that is, being able to tell where the original app is for step 3).

**Comment by duncanwilcox** [comment]
How about offering to download an update to ~/Downloads when running from a read only volume? With or without translocation, there's no notification that there's un update, and translocation makes this more frequent. Not a solution to automatic updates, but at least updates are shown.

**Comment by pornel** [comment]
I'm not so keen on this any more. 

Anything we do here will be pretty hacky, and if it will work at all, it's likely to be blocked by Apple later (since they may consider any workaround for translocation a security vulnerability and patch it).

Assumptions about app locations, folders, etc. can go wrong.

In the end it's perhaps best to just explain the issue to users, and/or switch to signed DMGs for app distribution.
